sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -0,0 +1,214 @@
+import com.google.cloud.bigquery.v3.ReadOptions.TableReadOptions;
+import com.google.cloud.bigquery.v3.TableReferenceProto;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import java.util.List;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryHelpers.TableRefToJson;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.ReadSessionOptions;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.options.ValueProvider.NestedValueProvider;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link BoundedSource} for reading BigQuery tables using the BigQuery parallel read API.
+ */
+class BigQueryParallelReadTableSource<T> extends BoundedSource<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReadTableSource.class);
+
+  private final BigQueryServices bqServices;
+  private final Coder<T> coder;
+  private final SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private final Session session;
+  private final ValueProvider<String> jsonTable;
+  private final ReadLocation initialReadLocation;
+  private final ReadSessionOptions readOptions;
+
+  private transient Long tableSizeBytes;
+  private transient List<BoundedSource<T>> cachedSplitResult;
+
+  private BigQueryParallelReadTableSource(
+      ValueProvider<String> jsonTable,
+      Session session,
+      ReadLocation readLocation,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readOptions) {
+    this.bqServices = checkNotNull(bqServices, "bqServices");
+    this.coder = checkNotNull(coder, "coder");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.jsonTable = checkNotNull(jsonTable, "jsonTable");
+    this.session = session;
+    this.initialReadLocation = readLocation;
+    this.readOptions = readOptions;
+  }
+
+  /**
+   * This method creates a {@link BigQueryParallelReadTableSource} with no initial session or read
+   * location.
+   */
+  public static <T> BigQueryParallelReadTableSource<T> create(
+      ValueProvider<TableReference> table,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readOptions) {
+    return new BigQueryParallelReadTableSource<>(
+        NestedValueProvider.of(checkNotNull(table, "table"), new TableRefToJson()),
+        null,
+        null,
+        parseFn,
+        coder,
+        checkNotNull(bqServices),
+        readOptions);
+  }
+
+  /*
+   * Gets the total size of the table.
+   */
+  @Override
+  public synchronized long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
```

Comment:
This should be the expected size produced by this reader not the total size.

Reply:
Just to be clear -- I think that Yiru took this implementation directly from the existing BigQuerySourceBase. In that case, however, split() returns a collection of AvroSource objects, and presumably these will return a better value for each individual source object.

We could approximate this behavior by computing the estimated table size once, at initial split() time, and by giving each child source object a fraction of the table based on the number of readers. This will not be accurate in the case where a failed reader is restarted, but it's a reasonable first-order approximation for the happy path. Does that sound reasonable?

Reply:
Yeah, that should be a reasonable first implementation.

Reply:
Done.